library/backend.cpp

```diff
@@ -1041,71 +869,461 @@ int dnet_cmd_bulk_read_new(struct dnet_net_state *st, struct dnet_cmd *cmd, void
+	value.AddMember("commands", commands_value, allocator);
+}
 
-	if (backend.has("cache")) {
-		const auto cache = backend["cache"];
-		cache_config = ioremap::cache::cache_config::parse(cache);
-	} else if (data->cache_config) {
-		cache_config = std::unique_ptr<ioremap::cache::cache_config>(new ioremap::cache::cache_config(*data->cache_config));
+void dnet_backend::statistics(uint64_t categories,
+                              rapidjson::Value &value,
+                              rapidjson::Document::AllocatorType &allocator) {
+	boost::shared_lock<boost::shared_mutex> guard(m_state_mutex);
+	fill_status(value, allocator);
+	if (categories & DNET_MONITOR_BACKEND)
+		fill_backend_stats(value, allocator);
+	if (categories & DNET_MONITOR_IO)
+		fill_io_stats(value, allocator);
+	if (categories & DNET_MONITOR_CACHE)
+		fill_cache_stats(value, allocator);
+	if (categories & DNET_MONITOR_COMMANDS)
+		fill_commands_stats(value, allocator);
+}
+
+dnet_backends_manager::dnet_backends_manager(struct dnet_node *node)
+: m_node(node) {
+	auto data = dnet_node_get_config_data(node);
+	m_backends.reserve(data->backends.size());
+
+	for (auto &config : data->backends) {
+		emplace(std::move(config));
 	}
 
-	io_thread_num = backend.at("io_thread_num", data->cfg_state.io_thread_num);
-	nonblocking_io_thread_num = backend.at("nonblocking_io_thread_num", data->cfg_state.nonblocking_io_thread_num);
+	// data->backends has been used and now longer needed, so clear it.
+	data->backends.clear();
+}
+
+dnet_backends_manager::~dnet_backends_manager() {
+	clear();
+}
+
+bool dnet_backends_manager::emplace(std::shared_ptr<backend_config> config) {
+	if (m_backends.find(config->backend_id) != m_backends.end())
+		return false;
+
+	m_backends.emplace(config->backend_id, std::make_shared<dnet_backend>(m_node, config));
+	return true;
+}
+
+std::shared_ptr<dnet_backend> dnet_backends_manager::get(uint32_t backend_id) {
+	boost::shared_lock<boost::shared_mutex> guard(m_backends_mutex);
+
+	auto it = m_backends.find(backend_id);
+	if (it == m_backends.end())
+		return nullptr;
+
+	return it->second;
+}
+
+int dnet_backends_manager::erase(uint32_t backend_id) {
+	boost::unique_lock<boost::shared_mutex> guard(m_backends_mutex);
+	return m_backends.erase(backend_id) == 0 ? -ENOENT : 0;
+}
+
+void dnet_backends_manager::clear() {
+	boost::upgrade_lock<boost::shared_mutex> guard(m_backends_mutex);
+	for (auto &item: m_backends) {
+		auto &backend = item.second;
+		backend->disable();
+	}
+
+	boost::upgrade_to_unique_lock<boost::shared_mutex> unique_guard(guard);
+	m_backends.clear();
+}
+
+int dnet_backends_manager::init_all(bool parallel) {
+	int err = 1;
+
+	if (parallel) {
+		try {
+			using namespace ioremap::elliptics;
+			session sess(m_node);
+			sess.set_filter(filters::all_with_ack);
+			sess.set_checker(checkers::no_check);
+			sess.set_exceptions_policy(session::no_exceptions);
+			sess.set_timeout(std::numeric_limits<unsigned>::max() / 2);
```

Comment:
This part of code enables all backends asynchronously. All generated commands are handled by nonblocking system threads which number is limited by config. Enabling of a backend is connected with reading of all blobs' indexes and filling internal structures which can take some time. This time is unpredictable at this point, so the timeout should be big enough to fit enabling of all backends. Since all generated commands will be placed directly to the system io pool's queue without sending via net, there is no need to expect and handle command loss. Therefore chosen timeout is so big, but why it is divided by 2 I don't know, probably we can remove this division.